lib/stm32/common/fdcan.c

```diff
@@ -0,0 +1,763 @@
+{
+	/* This piece of code may look wrong, after one examines
+	 * STM32G4 datasheet and/or g4/memorymap.h. There are three
+	 * memory regions defined for FDCANx_RAM_BASE. They are 0x400
+	 * bytes apart as per chapter 2.2.2 of [RM0440].
+	 *
+	 * It turns out, that these addresses are not in line with what
+	 * is specified later in chapter 44.3.3 of [RM0440]. There it is
+	 * stated, that message RAMs are packed and in case of multiple
+	 * FDCAN blocks, message RAM for n-th FDCAN starts at address
+	 * end of (n-1)-th block + 4 (explicitly, offset 0x354).
+	 *
+	 * It turns out, that this statement is also false! In fact FDCAN
+	 * message RAMs are packed tightly and n-th block starts immediately
+	 * after (n-1)-th block ends. Thus offset is going to be computed
+	 * using formula:
+	 *
+	 * FDCAN1_RAM_BASE + (block_id * sizeof(struct fdcan_message_ram))
+	 */
+	if (canport == CAN1) {
+		return (struct fdcan_message_ram *) (FDCAN1_RAM_BASE + 0);
+	} else if (canport == CAN2) {
+		return (struct fdcan_message_ram *)
+			(FDCAN1_RAM_BASE + sizeof(struct fdcan_message_ram));
+	} else if (canport == CAN3) {
+		return (struct fdcan_message_ram *)
+			(FDCAN1_RAM_BASE + (2 * sizeof(struct fdcan_message_ram)));
+	}
+
+	return (void *) 0;
+}
+
+/** Converts frame length to DLC value.
+ * Works for both CAN and FDCAN frame lengths. If length
+ * is invalid value, then returns 0xFF.
+ *
+ * @param length intended frame payload length in bytes
+ * @returns DLC value representing lengths or 0xFF if length cannot
+ * be encoded into DLC format (applies only to FDCAN frame lengths)
+ */
+static uint32_t fdcan_length_to_dlc(uint8_t length)
+{
+	if (length <= 8) {
+		return length;
+	} else if (length <= 24) {
+		if ((length % 4) != 0) {
+			return 0xFF;
+		}
+		return 8 + ((length - 8) / 4);
+	} else {
+		if ((length % 16) != 0) {
+			return 0xFF;
+		}
+		return 11 + (length / 16);
+	}
+}
+
+/** Converts DLC value into frame payload length.
+ * Works for both CAN and FDCAN DLC values.
+ *
+ * @param dlc DLC value
+ * @returns data payload length in bytes
+ */
+static uint8_t fdcan_dlc_to_length(uint32_t dlc)
+{
+	if (dlc <= 8) {
+		return dlc;
+	} else if (dlc <= 12) {
+		return 8 + ((dlc - 8) * 4);
+	} else {
+		return 16 + ((dlc - 12) * 16);
+	}
+}
+
+/* --- FD-CAN functions ----------------------------------------------------- */
+
+/** @ingroup fdcan_file */
+/**@{
+ * */
+
+/** Put FDCAN block into INIT mode for setup
+ * Initialize the selected CAN peripheral block. This function will switch CAN block
+ * into initialization mode and set crucial configuration flags and timing registers.
+ * CAN block is then left in initialization mode in order to
+ *
+ * @param[in] canport CAN register base address @ref can_reg_base.
+ * @param [in] timeout Amount of empty busy loops, which routine should wait for FDCAN
+ *				confirming that it entered INIT mode. If set to 0, function will return
+ *				immediately.
```

Comment:
what's the function that can be used to then check for it being ready? that should be linked here too.